File: trees/binaryTreePreorderTraversal.js
// https://leetcode.com/problems/binary-tree-preorder-traversal/
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// DFS
const preorderTraversal = (root) => {
  const stack = [root];
  const result = [];

  while (stack.length) {
    const node = stack.pop();
    if (node) {
      result.push(node.val);
      stack.push(node.right);
      stack.push(node.left);
    }
  }
  return result;
};

// Visited flag
const preorderTraversal = (root) => {
  const stack = [[root, false]];
  const res = [];

  while (stack.length) {
    const [node, visitedFlag] = stack.pop();
    if (node) {
      if (visitedFlag) {
        res.push(node.val);
      } else {
        stack.push([node.right, false]);
        stack.push([node.left, false]);
        stack.push([node, true]);
      }
    }
  }

  return res;
};

File: trees/binaryTreePreorderTraversal.py
# https://leetcode.com/problems/binary-tree-preorder-traversal/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # DFS
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        stack, res = [root], []
        while stack:
            node = stack.pop()
            if node:
                res.append(node.val)
                stack.append(node.right)
                stack.append(node.left)
        return res


class Solution:
    # Visited flag
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        traversal, stack = [], [(root, False)]
        while stack:
            node, visited = stack.pop()
            if node:
                if visited:
                    traversal.append(node.val)
                else:
                    stack.append((node.right, False))
                    stack.append((node.left, False))
                    stack.append((node, True))

        return traversal


def preorderTraversal1(self, root):
    # recursively
    res = []
    self.dfs(root, res)
    return res


def dfs(self, root, res):
    if root:
        res.append(root.val)
        self.dfs(root.left, res)
        self.dfs(root.right, res)
